### PR TITLE
Fix bug #1183 in DropDownList pre-scroll to last selected item is incorrect.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
There were several contributing problems all related to the BORDER_SIZE.
1. In the DropDownList::CorrectListSize() it was not accounting for the
borders of the list.
2. In ListBox BringRowIntoView() and FirstRowShownWhenBottomIs() both
were not accounting for the BORDER_SIZE.

### CODE REVIEW

File: GG/src/ListBox.cpp [hunk 1]
```diff
@@ -1173,6 +1173,7 @@ void ListBox::BringRowIntoView(iterator it)
 
     Y y_offset(BORDER_THICK), it_y_offset(BORDER_THICK);
     Y first_row_y_offset(BORDER_THICK), last_row_y_offset(BORDER_THICK);
+    iterator it_end_row = --m_rows.end();
```
Comment (geoffthemedio) [inline GG/src/ListBox.cpp]: is this safe in the event of an empty listbox?

Comment (LGM-Doyle) [reply GG/src/ListBox.cpp]: Yes.

The line above
````
    // m_first_row_shown only equals end() if the list is empty, hence 'it' is invalid.
    if (m_first_row_shown == m_rows.end())
        return;
````
should be checking an invariant of the ListBox that is preserved in all the insert/remove operations.  If this is not true, other parts of ListBox are broken.

**Comment by LGM-Doyle** [comment]
I think that this fixes the following things:
- The 4 pixel gap that you observed at the bottom of the drop down list.
- the ListBox would draw a full sized selection highlight if the selected item was partially scrolled out of the top of the ListBox.  Now it only highlights the visible portion of the selected item.
- the ListBox could scroll the list even after it determined that the vscroll was not needed.  Now it displays the entire list.
- the drop down list near the bottom of the app would be draw outside of the app area.  Now it draws as many items as it can in the space remaining, but never fewer than one.  This fixes issues #1196 and #969, if you are willing to accept that a drop down box with less that one row of space will still be drawn offscreen.  This is better than not drawing anything, because keyboard navigation still works.

File: GG/src/DropDownList.cpp [hunk 2]
```diff
@@ -109,6 +109,9 @@ class ModalListPicker : public Control
         list location not tracking the anchor.*/
     void WindowResizedSlot(X x, Y y);
 
+    /** Used by CorrectListSize() to determine the list height from the current first shown row. */
+    void DetermineListHeight(Pt drop_down_size);
```
Comment (geoffthemedio) [inline GG/src/DropDownList.cpp]: standard seems to be to pass as `const Pt&` rather than by value

File: GG/src/ListBox.cpp [hunk 3]
```diff
@@ -1159,55 +1159,58 @@ void ListBox::SetSelections(const SelectionSet& s, bool signal/* = false*/)
...
         return;
 
     // m_first_row_shown only equals end() if the list is empty, hence 'it' is invalid.
     if (m_first_row_shown == m_rows.end())
         return;
 
-    // Find the y offsets of the first and last shown rows and 'it'.
-    bool first_row_found(false), last_row_found(false), it_found(false);
+    // Find the y offsets of the first and last shown rows and target.
+    bool first_row_found(false), last_row_found(false), target_found(false);
```
Comment (geoffthemedio) [inline GG/src/ListBox.cpp]: @adrianbroher isn't fond of multiple definitions per line

**Comment by Vezzra** [comment]
> This fixes issues #1196 and #969, if you are willing to accept that a drop down box with less that one row of space will still be drawn offscreen.

I guess that's acceptable. So IMO merging this should completely resolve these issues.

**Pull request merged**